apollo-router/src/apollo_telemetry.rs

```diff
@@ -0,0 +1,590 @@
+                // XXX Since normalization is expensive, try to reduce the
+                // amount of normalization by doing an exact string match
+                // on a query. This might not save a lot of work and may
+                // result in too much caching, so re-visit this decision
+                // post-integration.
+                let key = self
+                    .normalized_queries
+                    .entry(query.as_str().to_string())
+                    .or_insert_with(|| normalize(operation_name, &query.as_str()));
+
+                // Retrieve DurationHistogram from our HashMap, or add a new one
+                let dh = dh_map
+                    .entry((client_name, client_version, key.clone()))
+                    .or_insert_with(|| DurationHistogram::new(None));
+                dh.increment_duration(span.end_time.duration_since(span.start_time).unwrap(), 1);
+            }
+        }
+
+        // Guarantee that the reporter is initialised
+        self.reporter
+            .get_or_try_init(|| async {
+                Reporter::try_new(self.collector.clone())
+                    .await
+                    .map_err::<ApolloError, _>(Into::into)
+            })
+            .await?;
+        let reporter = self.reporter.get_mut().unwrap();
+        // Convert the configuration data into a reportable form
+        let graph: ReporterGraph = self.graph.as_ref().unwrap().into();
+
+        // Report our consolidated statistics
+        for ((client_name, client_version, key), dh) in dh_map.into_iter() {
+            tracing::debug!("reporting entries: {}", dh.entries);
+            let stats = ContextualizedStats {
+                context: Some(StatsContext {
+                    client_name,
+                    client_version,
+                }),
+                query_latency_stats: Some(QueryLatencyStats {
+                    latency_count: dh.buckets,
+                    request_count: dh.entries,
+                    ..Default::default()
+                }),
+                ..Default::default()
+            };
+
+            let msg = reporter
+                .submit_stats(graph.clone(), key, stats)
+                .await
+                .map_err::<TraceError, _>(|e| e.to_string().into())?
+                .into_inner()
+                .message;
+            tracing::trace!("server response: {}", msg);
+        }
+
+        Ok(())
+    }
+}
+
+// Taken from TS implementation
+static GRAPHQL_PARSE_FAILURE: &str = "## GraphQLParseFailure\n";
+static GRAPHQL_VALIDATION_FAILURE: &str = "## GraphQLValidationFailure\n";
+static GRAPHQL_UNKNOWN_OPERATION_NAME: &str = "## GraphQLUnknownOperationName\n";
+
+fn normalize(op: Option<&opentelemetry::Value>, query: &str) -> String {
+    // If we don't have an operation name, we can't do anything useful
+    // with this query. Just return the appropriate error.
+    let op_name: String = match op {
```

Comment:
You are correct. I'll raise an issue to add a comment for this. Is this a problem that could be solved by enhancing the router_bridge interface as your previously described? (I think not, but I'm asking anyway).

Reply:
I think it could — IMO it doesn't really make sense to "resolve" the operation name for a request until you've validated the request (because then you can rely on uniqueness of operation names, etc).